compiler/rustc_typeck/src/check/expr.rs

```diff
@@ -1073,13 +1106,14 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             }
         });
 
+        let type_ascr_ctxt = expected.get_coercion_ctxt();
         let elt_ts_iter = elts.iter().enumerate().map(|(i, e)| match flds {
             Some(ref fs) if i < fs.len() => {
                 let ety = fs[i].expect_ty();
-                self.check_expr_coercable_to_type(&e, ety, None);
+                self.check_expr_coercable_to_type(&e, ety, None, type_ascr_ctxt);
                 ety
             }
-            _ => self.check_expr_with_expectation(&e, NoExpectation),
+            _ => self.check_expr_with_expectation(&e, NoExpectation(type_ascr_ctxt)),
```

Comment:
this should always be considered a coercion site, but we may not have a useful hint